src/metabase/db/connection.clj

```diff
@@ -35,3 +73,35 @@
     :postgres :ansi
     :h2       :h2
     :mysql    :mysql))
+
+;; TODO -- you can just use [[*application-db*]] directly, we can probably get rid of this and use that directly instead
+(defn data-source
+  "Get a data source for the application DB, derived from environment variables. This is NOT a pooled data source!
+  That's created later as part of [[metabase.db/setup-db!]] -- use [[toucan.db/connection]] if you want to get
+  a [[clojure.java.jdbc]] spec for the connection pool."
+  ^javax.sql.DataSource []
+  (:data-source *application-db*))
```

Comment:
would using the field access rather than keyword access actually be meaningful here?

Reply:
Yeah, it is slightly faster altho I'm not really sure I'd call it meaninful. It's a little less than a fifth of a nanosecond faster. It looks like the main difference is `.data-source` does direct field access while the keyword usage invokes the var which ends up doing an [`IKeywordLookup`](https://github.com/clojure/clojure/blob/b1b88dd25373a86e41310a525a21b497799dbbf2/src/jvm/clojure/lang/IKeywordLookup.java). It's worth noting that the lookup thunk is calculated at compile time and `(:data-source *application-db*)` is more or less the same as doing this

```clj
(def ^clojure.lang.ILookupThunk lookup-thunk
  (reify clojure.lang.ILookupThunk
    (get [this object]
      (if (identical? (class this) ApplicationDB)
        (.data-source ^ApplicationDB object)
        this))))

(let [app-db *application-db*
      result (.get thunk app-db)]
  (if (= result lookup-thunk) ; class was different
    (.get (recalculate-lookup-thunk app-db))
    result))
``` 

```clj
(reify
  clojure.lang.ILookupThunk
  (get [this object] 
    (if (identical? (class object) ApplicationDB) 
      (.data-source object)
      thunk)))
```

Here's profiling and disassembly (edited a bit for clarity)

### `(:data-source *application-db*)`

```
Evaluation count : 7974308400 in 60 samples of 132905140 calls.
             Execution time mean : 3.167893 ns
    Execution time std-deviation : 0.212715 ns
   Execution time lower quantile : 2.959563 ns ( 2.5%)
   Execution time upper quantile : 3.530544 ns (97.5%)
                   Overhead used : 4.065358 ns
```

```java
public final class fn__54032 extends AFunction
{
    public static final Var var;
    static final KeywordLookupSite site;
    static ILookupThunk thunk;
    
    public static Object invokeStatic() {
        final Object v = var.get();
        Object o2;
        if (thunk == (o2 = thunk.get(v))) {
            o2 = (thunk = site.fault(v)).get(v);
        }
        return o2;
    }
    
    static {
        var = RT.var("metabase.db.connection", "*application-db*");
        thunk = (site = new KeywordLookupSite(RT.keyword(null, "data-source")));
    }
}
```

### `(.data-source *application-db*)`

```
Evaluation count : 8502007740 in 60 samples of 141700129 calls.
             Execution time mean : 2.999667 ns
    Execution time std-deviation : 0.015575 ns
   Execution time lower quantile : 2.976435 ns ( 2.5%)
   Execution time upper quantile : 3.031843 ns (97.5%)
                   Overhead used : 4.065358 ns
```

```java
public final class fn__54028 extends AFunction
{
    public static final Var var;

    public static Object invokeStatic() {
        final Object v = var.get();
        return ((ApplicationDB)v).data_source;
    }

    static {
        var = RT.var("metabase.db.connection", "*application-db*");
    }
}
```

Reply:
I'll change it anyway since it does result in smaller code alto this isn't something that I would go out of my way to worry about except in super tight performance-critical loops

Reply:
changed

Reply:
I am not going to give a full lecture on sampling statistics but it's not possible to tell differences of nanoseconds on 60 samples on commodity hardware